### PR TITLE
Add deprecation warnings to results classes and methods

### PR DESCRIPTION
**Pull request by desertaxle**
This PR adds deprecation warning to following because they have been superseded by new results mechanics:
- `ResultStore.create_result`
- `BaseResult`
- `PersistedResult`

This PR also updates our testing logic to fail whenever a `PrefectDeprecationWarning` is raised. A fixture `ignore_prefect_deprecation_warnings` can suppress failures until the deprecation period has passed. Once the deprecation period has passed, the warnings will cause test failures again, prompting us to remove the deprecated code.

Related to https://github.com/PrefectHQ/prefect/issues/15208

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [ ] If this pull request adds new functionality, it includes unit tests that cover the changes
- [ ] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [ ] If this pull request adds functions or classes, it includes helpful docstrings.

### CODE REVIEW

File: src/prefect/results.py [hunk 1]
```diff
@@ -755,6 +756,11 @@ async def await_for_lock(self, key: str, timeout: Optional[float] = None) -> boo
             )
         return await self.lock_manager.await_for_lock(key, timeout)
 
+    @deprecated.deprecated_callable(
+        start_date="Sep 2024",
+        end_date="Nov 2024",
+        help="Use `create_result_record` instead.",
+    )
     @sync_compatible
     async def create_result(
```
Comment (desertaxle) [inline src/prefect/results.py:765]: Deprecation warning for `create_reuslt`

File: src/prefect/results.py [hunk 2]
```diff
@@ -1078,6 +1103,9 @@ def __dispatch_key__(cls, **kwargs):
         return cls.__name__ if isinstance(default, PydanticUndefinedType) else default
 
 
+@deprecated.deprecated_class(
+    start_date="Sep 2024", end_date="Nov 2024", help="Use `ResultRecord` instead."
+)
 class PersistedResult(BaseResult):
```
Comment (desertaxle) [inline src/prefect/results.py:1109]: Deprecation warning for `PersistedResult`

File: src/prefect/results.py [hunk 3]
```diff
@@ -1027,7 +1044,15 @@ def deserialize_from_result_and_metadata(
...
+        if not isinstance(other, ResultRecord):
+            return False
+        return self.metadata == other.metadata and self.result == other.result
+
 
+@deprecated.deprecated_class(
+    start_date="Sep 2024", end_date="Nov 2024", help="Use `ResultRecord` instead."
+)
 @register_base_type
 class BaseResult(BaseModel, abc.ABC, Generic[R]):
```
Comment (desertaxle) [inline src/prefect/results.py:1057]: Deprecation warning for `BaseResult`

File: tests/fixtures/deprecation.py [hunk 4]
```diff
@@ -5,18 +5,47 @@
...
+        )
+        # Parse the date
+        deprecation_date = datetime.strptime(date_str, "%b %Y").date().replace(day=1)
+
+        # Check if the current date is after the start of the month following the deprecation date
+        current_date = datetime.now().date().replace(day=1)
+        return current_date > deprecation_date
+    except Exception:
+        # Reraise in cases of failure
+        return True
```
Comment (desertaxle) [inline tests/fixtures/deprecation.py:36]: I added this check to know when the deprecated code is ready to be removed. As a result, code path that raise deprecation warnings in tests will need to use the `ignore_prefect_deprecation_warnings` fixture directly.

**Review by cicdw — APPROVED** [review]

**Pull request merged**